src/main/java/com/linkedin/xinfra/monitor/services/MultiClusterTopicManagementService.java

```diff
@@ -415,90 +414,103 @@ int numPartitions() throws InterruptedException, ExecutionException {
-            .isEmpty()) {
-          LOGGER.info(
-              "MultiClusterTopicManagementService will increase the replication factor of the topic {} in cluster"
-                  + "from {} to {}", _topic, currentReplicationFactor, expectedReplicationFactor);
-          reassignPartitions(_adminClient, brokers, _topic, partitionInfoList.size(), expectedReplicationFactor);
+      if (_replicationFactor < currentReplicationFactor) {
+        LOGGER.debug(
+            "Configured replication factor {} is smaller than the current replication factor {} of the topic {} in cluster.",
+            _replicationFactor, currentReplicationFactor, _topic);
+      }
 
-          partitionReassigned = true;
-        }
+      if (expectedReplicationFactor > currentReplicationFactor && Utils.ongoingPartitionReassignments(_adminClient)
+          .isEmpty()) {
+        LOGGER.info("MultiClusterTopicManagementService will increase the replication factor of the topic {} in cluster"
+            + "from {} to {}", _topic, currentReplicationFactor, expectedReplicationFactor);
+        reassignPartitions(_adminClient, brokers, _topic, partitionInfoList.size(), expectedReplicationFactor);
+        partitionReassigned = true;
+      }
 
-        // Update the properties of the monitor topic if any config is different from the user-specified config
-        Properties currentProperties = zkClient.getEntityConfigs(ConfigType.Topic(), _topic);
-        Properties expectedProperties = new Properties();
-        for (Object key : currentProperties.keySet()) {
-          expectedProperties.put(key, currentProperties.get(key));
-        }
-        for (Object key : _topicProperties.keySet()) {
-          expectedProperties.put(key, _topicProperties.get(key));
-        }
+      // Update the properties of the monitor topic if any config is different from the user-specified config
+      Properties currentProperties = new Properties(); // ConfigResource -> Collection<AlterConfigOp>
+      Properties expectedProperties = new Properties(); // ConfigResource -> Collection<AlterConfigOp>
+      DescribeConfigsResult describeConfigsResult =
+          _adminClient.describeConfigs(Collections.singleton(new ConfigResource(ConfigResource.Type.TOPIC, _topic)));
 
-        if (!currentProperties.equals(expectedProperties)) {
-          LOGGER.info("MultiClusterTopicManagementService will overwrite properties of the topic {} "
-              + "in cluster from {} to {}.", _topic, currentProperties, expectedProperties);
-          zkClient.setOrCreateEntityConfigs(ConfigType.Topic(), _topic, expectedProperties);
-        }
+      for (Map.Entry<ConfigResource, KafkaFuture<Config>> entry : describeConfigsResult.values().entrySet()) {
+        currentProperties.put(entry.getKey(),
+            new AlterConfigOp(entry.getValue().get().get(_topic), AlterConfigOp.OpType.SET));
```

Comment:
Is this right?
From what I understand from the API, the `describeConfigsResult.values().entrySet()` will have only 1 entry with the topic ConfigResource as key, so then `entry.getValue().get().get(_topic)` will return `null` as the keys for `Config` object are configuration keys, no?

Reply:
AlterConfigOp API: 
`public AlterConfigOp(ConfigEntry configEntry, OpType operationType)`

ConfigResource configResource = entry.getKey();
Config config = entry.getValue().get();
ConfigEntry configEntry = config.get(_topic);

What are you suggesting otherwise for the value of 
`currentProperties.put(configResource, new AlterConfigOp(?,?));`
?

Reply:
The issue is not on the `AlterConfigOp` part, it's on the `describeConfigs` part and the way it's parsed.

To explain myself more easily:
```java
ConfigResource topicConfigResource = new ConfigResource(ConfigResource.Type.TOPIC, _topic);
Config currentConfig = _adminClient.describeConfigs(Collections.singleton(topicConfigResource)).all().get().get(topicConfigResource);
```
will give you this value for `currentConfig`:
```
Config(entries=[
    ConfigEntry(name=compression.type, value=producer, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=leader.replication.throttled.replicas, value=, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=message.downconversion.enable, value=true, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=min.insync.replicas, value=2, source=STATIC_BROKER_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=segment.jitter.ms, value=0, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=cleanup.policy, value=delete, source=DYNAMIC_TOPIC_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=flush.ms, value=9223372036854775807, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=follower.replication.throttled.replicas, value=, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=segment.bytes, value=536870912, source=DYNAMIC_TOPIC_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=retention.ms, value=3700000, source=DYNAMIC_TOPIC_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=flush.messages, value=9223372036854775807, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=message.audit.enable, value=false, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=message.format.version, value=2.3-IV1, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=file.delete.delay.ms, value=60000, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=max.compaction.lag.ms, value=9223372036854775807, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=max.message.bytes, value=1000012, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=min.compaction.lag.ms, value=0, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=message.timestamp.type, value=CreateTime, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=preallocate, value=false, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=min.cleanable.dirty.ratio, value=0.5, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=index.interval.bytes, value=4096, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=unclean.leader.election.enable, value=false, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=retention.bytes, value=1073741824, source=DYNAMIC_TOPIC_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=delete.retention.ms, value=86400000, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=segment.ms, value=604800000, source=DYNAMIC_TOPIC_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=message.timestamp.difference.max.ms, value=9223372036854775807, source=DEFAULT_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[]),
    ConfigEntry(name=segment.index.bytes, value=10485760, source=DYNAMIC_TOPIC_CONFIG, isSensitive=false, isReadOnly=false, synonyms=[])
])
```

So there is no `_topic` key in `config`.

I was trying to achieve the same goal before seeing your PR, and I came to this: https://github.com/linkedin/kafka-monitor/commit/444b25cf6e8eb8bbcb4544688e6589d104829e68

Reply:
I have updated the PR to reflect the iteration over Config entires of the entry of `describeConfigResult`. Would you take a look at the change again?

Reply:
`currentProperties` will only contains 1 entry with the last ConfigEntry as it's putting always with the same unique key `configResource`.

I made the PR #319 that does the trick by overwriting the changed configs if necessary.
I tested it and if the `retention.ms` is changed from the expected value from the xinfra-monitor.properties, then it's set again to the right value.